### DISCUSSION TITLE
cargo-fetcher proposal on bitbake list

### DISCUSSION THREAD

**Discussion by rwmacleod** [meta-rust/meta-rust#315] [Q&A]
Please review the proposed cargo-fetcher work submitted to the bitbake list:
  https://lists.openembedded.org/g/bitbake-devel/message/12022?p=,,,20,0,0,0::Created,,cargo,20,2,0,80454027

I think this would break lots of recipes that use the meta-rust layer.
Alternative proposals that work for recipes using the work of meta-rust would be appreciated on the bitbake list.

**Comment by janderholm** [comment]
Since you asked for alternative approaches, here's a suggestion I have. It works and has some merits such as being able to properly list the licenses of each dependency. 

It goes like this:

For a given package, parse Cargo.lock for all dependencies. ("cargo metadata" is useful) 

Generate one recipe for each dependency.

Each recipe downloads the crate source from crates.io (or wherever).

Install the source into /usr/src/vendored-sources/cratename-version

In the main recipe, list all crates as dependencies.

Set the vendoring directory to /usr/src/vendored-sources.


The main advantage of this approach is all cargo packages will share dependencies, and each source recipe can list the license and point to a license file like is usually done. It might even be possible to patch the source, although I am unsure how it will work with the locking of cargo. 

The main disadvantage is there will be loads of recipes since a cargo package typically use a lot of dependencies. They need to be cleaned now and then, and there's potential for lots of duplicated recipes in different layers. 

I can put together a small example if someone would like to see it in action.
